### PR TITLE
 crypto: add certificate request support

### PR DESCRIPTION
**Pull request by aberaud**

### CODE REVIEW

File: src/crypto.cpp [hunk 1]
```diff
@@ -572,6 +572,153 @@ PublicKey::getLongId() const
...
+        return NameType::URI;
+    case GNUTLS_SAN_IPADDRESS:
+        return NameType::IP;
+    default:
+        return NameType::UNKNOWN;
+    }
+}
+
+static gnutls_x509_subject_alt_name_t
+typeFromName(NameType type)
```
Comment (AmarOk1412) [inline src/crypto.cpp]: `typeFromGnuTLS` <=> `GnuTLSFromName` maybe here?

The previous function is the invert operation, should be reflected into the API

File: src/crypto.cpp [hunk 2]
```diff
@@ -572,6 +572,153 @@ PublicKey::getLongId() const
...
+CertificateRequest::sign(const PrivateKey& key, const std::string& password)
+{
+    gnutls_x509_crq_set_version(request, 1);
+    if (not password.empty())
+        gnutls_x509_crq_set_challenge_password(request, password.c_str());
+
+    if (auto err = gnutls_x509_crq_set_key(request,  key.x509_key))
+        throw CryptoException(std::string("Can't set certificate request key: ") + gnutls_strerror(err));
+        
+    if (auto err = gnutls_x509_crq_sign(request, key.x509_key))
```
Comment (AmarOk1412) [inline src/crypto.cpp]: Seems deprecated. `gnutls_x509_crq_privkey_sign` instead.

Source https://manpages.debian.org/stretch/gnutls-doc/gnutls_x509_crq_sign.3.en.html

File: src/crypto.cpp (hunk 1 above)
Comment (aberaud) [reply src/crypto.cpp]: done

File: src/crypto.cpp (hunk 2 above)
Comment (aberaud) [reply src/crypto.cpp]: done

**Comment by AmarOk1412** [comment]
Ok for me!

**Pull request merged**